feat: ColorPicker support gradient color

zombieJ opened the pull request:
<!--
First of all, thank you for your contribution! 😄
For requesting to pull a new feature or bugfix, please send it from a feature/bugfix branch based on the `master` branch.
Before submitting your pull request, please make sure the checklist below is confirmed.
Your pull requests will be merged after one of the collaborators approve.
Thank you!
-->

[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

close #50004
close https://github.com/ant-design/ant-design/issues/42523

### 💡 Background and solution

<!--
1. Describe the problem and the scenario.
2. GIF or snapshot should be provided if includes UI/interactive modification.
3. How to fix the problem, and list the final API implementation and usage sample if that is a new feature.
-->

### 📝 Changelog

<!--
Describe changes from the user side, and list all potential break changes or other risks.
--->

- Use a developer-oriented tone and narrative style.
- Describe the user's first-hand experience of the issue and its impact on developers, rather than your solution approach.
- Refer to: https://ant.design/changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |      ColorPicker support gradient color type. Fix controlled mode not working.     |
| 🇨🇳 Chinese |   ColorPicker 支持渐变色选择。修复受控模式不生效的问题。      |

zombieJ committed:
refactor: support type update

zombieJ committed:
chore: update clear style

zombieJ committed:
Merge remote-tracking branch 'origin/feature' into color-multiple

zombieJ committed:
Merge remote-tracking branch 'origin/feature' into color-multiple

zombieJ committed:
chore: bump color picker

zombieJ committed:
chore: use slider

zombieJ committed:
chore: bump color picker

zombieJ committed:
Merge remote-tracking branch 'origin/feature' into color-multiple

zombieJ committed:
chore: range slider

zombieJ committed:
chore: layout

zombieJ committed:
chore: useModeColor

zombieJ committed:
chore: simplify

zombieJ committed:
chore: bump color picker

zombieJ committed:
refactor: event

zombieJ committed:
chore: tmp lock check

zombieJ committed:
chore: of it

zombieJ committed:
chore: update ts def

zombieJ committed:
chore: update ts def

zombieJ committed:
chore: remove useless ts

zombieJ committed:
chore: linear

zombieJ committed:
chore: adjust style

zombieJ committed:
chore: rm useless code

zombieJ committed:
chore: fill color

zombieJ committed:
chore: basic linear

zombieJ committed:
chore: support toStr

zombieJ committed:
chore: limit minCount

zombieJ committed:
chore: use cache

zombieJ committed:
chore: drag support:

zombieJ committed:
chore: yes

zombieJ committed:
chore: update demo

zombieJ committed:
chore: useLayoutEffect instead

zombieJ committed:
chore: fix click to add point

zombieJ committed:
chore: add smmoth

zombieJ committed:
chore: support of locale

zombieJ committed:
chore: add locale

zombieJ committed:
chore: fix lint

zombieJ committed:
chore: adjust style

li-jia-nan commented on components/color-picker/style/index.ts:
```diff
@@ -75,12 +75,12 @@ const genClearStyle = (
         width: 40, // maximum
         height: 2, // fixed
-        transformOrigin: 'right',
+        transformOrigin: `calc(100% - 1px) 1px`,
```
这里不用 token 吗？

zombieJ replied:
看看上面一行 😈

zombieJ committed:
chore: fix lint

zombieJ committed:
chore: fix style

zombieJ committed:
test: fix test case

zombieJ committed:
chore: fix popover

li-jia-nan commented on components/color-picker/__tests__/index.test.tsx:
```diff
@@ -592,9 +594,13 @@ describe('ColorPicker', () => {
       );
     };
     const { container } = render(<Demo />);
+    console.log('disable');
```
console

zombieJ committed:
test: fix test case

zombieJ committed:
chore: fix test

zombieJ committed:
test: clean up

zombieJ committed:
chore: fix lint

zombieJ commented:
@SocketSecurity ignore npm/@ant-design/tools@18.0.2
@SocketSecurity ignore npm/sharp@0.33.4
@SocketSecurity ignore npm/@biomejs/biome@1.8.3

zombieJ committed:
chore: fix lint

zombieJ committed:
chore: fix lint

zombieJ committed:
test: coverage

zombieJ committed:
test: coverage

zombieJ committed:
test: coverage

zombieJ committed:
test: coverage

zombieJ committed:
test: coverage

li-jia-nan commented:
大了 1.77kb

zombieJ committed:
test: coverage

zombieJ committed:
chore: fix docs

zombieJ commented:
> 大了 1.77kb

多了一个 fast-color，全部替换掉后会还回去 2KB

zombieJ committed:
chore: merge feature

afc163 commented:
为啥没有 preview 和 size-limit 链接

afc163 commented on components/color-picker/demo/line-gradient.md:
```diff
@@ -0,0 +1,7 @@
+
+## en-US
+
+Basic Usage.
```
这个文案不太对。

afc163 reviewed (changes requested):
1. 渐变色演示的描述不太对。
2. 添加一个节点后，如何删除掉？
3. 点击节点白色外圈时不会选择节点，会创建一个新节点。这里要多给点容错度，不然很容易在相近位置点好几个出来。
4. input 内的 selection 效果还需不需要做？
5. 好像不支持角度？
6. 键盘操作好像也不太行。

zombieJ committed:
docs: update demo desc

zombieJ committed:
chore: enhance hover range

zombieJ committed:
fix: delete not working

zombieJ committed:
chore: fix lint

zombieJ commented:
#### 渐变色演示的描述不太对。

Updated

#### 添加一个节点后，如何删除掉？

拖拽到外部，或者删除按键（出了 BUG 删不掉，已经修复并添加测试）

#### 点击节点白色外圈时不会选择节点，会创建一个新节点。这里要多给点容错度，不然很容易在相近位置点好几个出来。

Updated

#### input 内的 selection 效果还需不需要做？

Updated，编辑渐变色同步高亮文本。

#### 好像不支持角度？

嗯，这一期不支持。

#### 键盘操作好像也不太行。

同上，fixed

zombieJ committed:
test: coverage

zombieJ committed:
test: coverage

li-jia-nan commented on components/color-picker/__tests__/gradient.test.tsx:
```diff
@@ -240,7 +242,8 @@ describe('ColorPicker.gradient', () => {
     );
 
-    // Drag remove first
+    console.log('~~~~~~~~~~~~');
```
console 没删大佬

zombieJ committed:
chore: clean up

afc163 commented:
<img width="340" alt="图片" src="https://github.com/user-attachments/assets/154c5e86-f4c0-4c2e-a50e-8e14794a6031">

左边应该垂直居中不？

li-jia-nan commented:
> 左边应该垂直居中不？

我感觉没必要，如果很高的话，左边的色块就在很下面了，不过感觉可以问问设计师的意见？

wanpan11 commented:
感觉这个当前 选中的节点 不太明显 😂

<img width="491" alt="image" src="https://github.com/user-attachments/assets/f9b402de-f815-4830-bd32-d874c0ff71fb">

li-jia-nan commented on components/slider/style/Context.ts:
```diff
@@ -0,0 +1,12 @@
+
+export interface SliderInternalContextProps {
+  handleRender?: RcSliderProps['handleRender'];
+  direction?: 'ltr' | 'rtl';
```
```suggestion
  direction?: DirectionType;
```

zombieJ commented:
> 左边应该垂直居中不？

原本是垂直居中的，切多行时不稳定感会加深丑的感觉。所以固定在上面位置了。

zombieJ committed:
chore: adjust

MadCcc commented:
1. 色相和透明度的拖动条感觉不需要点击之后放大节点（不如说感觉默认状态下的 handler 太小了）。如果要展示 focus 状态的话建议和 slider 一致，hover 的时候就放大。
2. 点击清除颜色后再选择颜色，渐变色条颜色不对。个人意见渐变色从无色恢复到有色时是不是默认添加两个节点比较好？
![image](https://github.com/user-attachments/assets/f66f14d2-1dfc-4122-b878-26156a3028a1)
3. 清除颜色的按钮布局在渐变色与单色下都有点难看，左边空了一大块，渐变色下更加明显了。

zombieJ committed:
chore: highlight

zombieJ committed:
chore: merge feature

zombieJ committed:
chore: adjust style

zombieJ committed:
chore: fix lint

zombieJ committed:
chore: update demo

zombieJ committed:
chore: memo perf

zombieJ committed:
refactor: up to down colors

zombieJ committed:
test: update snapshot

zombieJ commented:
@SocketSecurity ignore npm/esbuild@0.18.20
@SocketSecurity ignore npm/@swc/core@1.4.2

afc163 commented:
弹出动画貌似不太对，应该从左上角位置弹出。

> 当然最好从箭头位置弹出。

zombieJ commented:
> 弹出动画貌似不太对，应该从左上角位置弹出。

业界的 ColorPicker 默认都是左下弹出，因为人们习惯性不会给 ColorPicker 留上方空间来进行选择。这和 Select 类组件逻辑是类似的，优先下方，不够再翻转。

<img width="280" alt="截屏2024-07-29 15 38 00" src="https://github.com/user-attachments/assets/e7979078-91bc-490b-97c2-0402fc44cab5">


<img width="364" alt="截屏2024-07-29 15 37 07" src="https://github.com/user-attachments/assets/e29bfc6d-cee4-493f-83ca-98855db22e11">

<img width="402" alt="截屏2024-07-29 15 42 10" src="https://github.com/user-attachments/assets/a8b2b28e-b8e7-4a0f-aa22-7f9ceb765a50">


弹出中心不对看起来是 Popover 的问题，可以在 master 里开新的 Patch 修一下~

zombieJ commented:
> 色相和透明度的拖动条...

Updated

> 点击清除颜色后再选择颜色...

Fixed

> 清除颜色的按钮布局在渐变色与单色下都有点难看，左边空了一大块，渐变色下更加明显了。

单选之前也是这样，暂时先不改这个布局了。

afc163 commented:
> 弹出动画貌似不太对，应该从左上角位置弹出。

我意思是弹出动画的方向，应从箭头位置为起点弹出。

afc163 commented:
https://github.com/ant-design/ant-design/pull/50028 这个还需要么

afc163 commented:
<img width="318" alt="图片" src="https://github.com/user-attachments/assets/d16d7082-2dda-4215-9600-997b406ec7fb">

选中之后按左右键，是不是可以改变百分比？

zombieJ commented:
> 选中之后按左右键，是不是可以改变百分比？现在没有这个交互

和 删除、添加 节点的事件时机有冲突，渐变色的键盘左右先去掉了

mufeng889 commented:
大佬   能让token支持渐变色吗？

diqye commented:
无法指定渐变方向。
